pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,336 @@
+	syncMetricsCache *syncutil.MetricsCache
+	tracker          *readiness.Tracker
+	processExcluder  *process.Excluder
+	registrar        *watch.Registrar
+	watchedSet       *watch.Set
+	gvkAggregator    *aggregator.GVKAgreggator
+	gvksToRemove     *watch.Set
+	gvksToSync       *watch.Set
+	replayErrChan    chan error
+	replayTicker     time.Ticker
+	reader           client.Reader
+}
+
+func NewCacheManager(config *CacheManagerConfig) (*CacheManager, error) {
+	if config.WatchedSet == nil {
+		return nil, fmt.Errorf("watchedSet must be non-nil")
+	}
+	if config.Registrar == nil {
+		return nil, fmt.Errorf("registrar must be non-nil")
+	}
+	if config.ProcessExcluder == nil {
+		return nil, fmt.Errorf("processExcluder must be non-nil")
+	}
+	if config.Tracker == nil {
+		return nil, fmt.Errorf("tracker must be non-nil")
+	}
+	if config.Reader == nil {
+		return nil, fmt.Errorf("reader must be non-nil")
+	}
+
+	cm := &CacheManager{
+		opa:              config.Opa,
+		syncMetricsCache: config.SyncMetricsCache,
+		tracker:          config.Tracker,
+		processExcluder:  config.ProcessExcluder,
+		registrar:        config.Registrar,
+		watchedSet:       config.WatchedSet,
+		reader:           config.Reader,
+	}
+
+	cm.gvkAggregator = aggregator.NewGVKAggregator()
+	cm.gvksToRemove = watch.NewSet()
+	cm.gvksToSync = watch.NewSet()
+
+	cm.replayTicker = *time.NewTicker(3 * time.Second)
+
+	return cm, nil
+}
+
+func (c *CacheManager) Start(ctx context.Context) error {
+	go c.updateDatastore(ctx)
+
+	<-ctx.Done()
+	return nil
+}
+
+// WatchGVKsToSync adjusts the watched set of gvks according to the newGVKs passed in
+// for a given {syncSourceType, syncSourceName}.
+func (c *CacheManager) WatchGVKsToSync(ctx context.Context, newGVKs []schema.GroupVersionKind, newExcluder *process.Excluder, syncSourceType, syncSourceName string) error {
+	netNewGVKs := []schema.GroupVersionKind{}
+	for _, gvk := range newGVKs {
+		if !c.gvkAggregator.IsPresent(gvk) {
+			netNewGVKs = append(netNewGVKs, gvk)
+		}
+	}
+	// mark these gvks for the background goroutine to sync
+	c.gvksToSync.Add(netNewGVKs...)
+
+	opKey := aggregator.Key{Source: syncSourceType, ID: syncSourceName}
+	currentGVKsForKey := c.gvkAggregator.List(opKey)
+
+	if len(newGVKs) == 0 {
+		// we are not syncing anything for this key anymore
+		if err := c.gvkAggregator.Remove(opKey); err != nil {
+			return fmt.Errorf("internal error removing gvks for aggregation: %w", err)
+		}
+	} else {
+		if err := c.gvkAggregator.Upsert(opKey, newGVKs); err != nil {
+			return fmt.Errorf("internal error upserting gvks for aggregation: %w", err)
+		}
+	}
+
+	// stage the new watch set for events for the sync_controller to be
+	// the current watch set ... [1/3]
+	newGvkWatchSet := watch.NewSet()
+	newGvkWatchSet.AddSet(c.watchedSet)
+	// ... plus the net new gvks we are adding ... [2/3]
+	newGvkWatchSet.Add(netNewGVKs...)
+
+	gvksToDeleteCandidates := getGVKsToDeleteCandidates(newGVKs, currentGVKsForKey)
```

Comment:
What cleanup to do in the cache should be decoupled from the updating of the cache specification (the GVKs and/or process excluder)